filters/src/voxel_grid.cpp

```diff
@@ -436,93 +446,65 @@ pcl::VoxelGrid<pcl::PCLPointCloud2>::applyFilter (PCLPointCloud2 &output)
     xyz_offset = Array4size_t (0, 4, 8, 12);
 
   index=0;
-  Eigen::VectorXf centroid = Eigen::VectorXf::Zero (centroid_size);
   Eigen::VectorXf temporary = Eigen::VectorXf::Zero (centroid_size);
 
-  for (std::size_t cp = 0; cp < index_vector.size ();)
+  for (const auto &cp : first_and_last_indices_vector)
   {
-    std::size_t point_offset = index_vector[cp].cloud_point_index * input_->point_step;
-    // Do we need to process all the fields?
-    if (!downsample_all_data_) 
-    {
-      memcpy (&pt[0], &input_->data[point_offset+input_->fields[x_idx_].offset], sizeof (float));
-      memcpy (&pt[1], &input_->data[point_offset+input_->fields[y_idx_].offset], sizeof (float));
-      memcpy (&pt[2], &input_->data[point_offset+input_->fields[z_idx_].offset], sizeof (float));
-      centroid[0] = pt[0];
-      centroid[1] = pt[1];
-      centroid[2] = pt[2];
-      centroid[3] = 0;
-    }
-    else
-    {
-      // ---[ RGB special case
-      // fill extra r/g/b centroid field
-      if (rgba_index >= 0)
-      {
-        pcl::RGB rgb;
-        memcpy (&rgb, &input_->data[point_offset + input_->fields[rgba_index].offset], sizeof (RGB));
-        centroid[centroid_size-4] = rgb.r;
-        centroid[centroid_size-3] = rgb.g;
-        centroid[centroid_size-2] = rgb.b;
-        centroid[centroid_size-1] = rgb.a;
-      }
-      // Copy all the fields
-      for (std::size_t d = 0; d < input_->fields.size (); ++d)
-        memcpy (&centroid[d], &input_->data[point_offset + input_->fields[d].offset], field_sizes_[d]);
-    }
+    // calculate centroid - sum values from all input points, that have the same idx value in index_vector array
+    unsigned int first_index = cp.first;
+    unsigned int last_index = cp.second;
 
-    std::size_t i = cp + 1;
-    while (i < index_vector.size () && index_vector[i].idx == index_vector[cp].idx) 
+    // index is centroid final position in resulting PointCloud
+    if (save_leaf_layout_)
+      leaf_layout_[index_vector[first_index].idx] = index;
+
+    //Limit downsampling to coords
+    if (!downsample_all_data_)
```

Comment:
Feels like we can set the upper limit on a loop variable to 3 or the num_fields and integrate the two loops.

Though I don't know how badly it'd harm performance for Vector4f's case

Reply:
To make sure I understand, you are suggesting putting these lines
```
        memcpy (&pt[0], &input_->data[point_offset+input_->fields[x_idx_].offset], sizeof (float));
        memcpy (&pt[1], &input_->data[point_offset+input_->fields[y_idx_].offset], sizeof (float));
        memcpy (&pt[2], &input_->data[point_offset+input_->fields[z_idx_].offset], sizeof (float));
        centroid[0] += pt[0];
        centroid[1] += pt[1];
        centroid[2] += pt[2];
```
Inside a for loop over the 0-2?  I could, I would need to cache the x_idx_,y_idx_,z_idx_ in a vector to access with the iterator.

Reply:
You can just use the same logic for offsets used in the else section, right?

Reply:
It was slightly different in the else section because the order of the fields didn't matter when it was doing the for loop over the field size (for example centroid[0] might not have corresponded with x, but it was put in the output in the location corresponding with the field[0] offset).  I put the if section in a for loop by making an index_t structure with the {x_ind_, y_ind_, z_ind_} so I could loop over 0,1,2 and have it correspond with x,y,z